File: solutions/cpp/323/solution.cpp
class Solution {
public:
    unordered_map<int, vector<int>> graph;
    vector<bool> seen;
    int countComponents(int n, vector<vector<int>>& edges) {
        seen = vector<bool>(n);
        // convert array of edges to list of adjacency
        for (const auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        int ans = 0;
        for (int i = 0; i < n; i++) {
            if (!seen[i]) {
                seen[i] = true;
                ans++;
                dfs(i);
            }
        }
        return ans;
    }

    void dfs(int node) {
        for (const auto& neighbor : graph[node]) {
            if (!seen[neighbor]) {
                seen[neighbor] = true;
                dfs(neighbor);
            }
        }
    }
};

File: solutions/golang/323/solution.go
func countComponents(n int, edges [][]int) int {
    seen := make([]bool, n)
    graph := make(map[int][]int)
    for _, edge := range edges {
        to, from := edge[0], edge[1]
        graph[to] = append(graph[to], from)
        graph[from] = append(graph[from], to)
    }
    ans := 0
    for i := 0; i < n; i++ {
        if !seen[i] {
            seen[i] = true
            ans++
            dfs(i, seen, graph)
        }
    }
    return ans
}

func dfs(node int, seen []bool, graph map[int][]int) {
    for _, neighbor := range graph[node] {
        if !seen[neighbor] {
            seen[neighbor] = true
            dfs(neighbor, seen, graph)
        }
    }
}
